examples/LidarISAM2Example.cpp

```diff
@@ -0,0 +1,518 @@
+    float roll;
+    float pitch;
+    float yaw;
+    double time;
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW   // make sure our new allocators are aligned
+} EIGEN_ALIGN16;                    // enforce SSE padding for correct memory alignment
+
+POINT_CLOUD_REGISTER_POINT_STRUCT (PointXYZIRPYT,
+                                   (float, x, x) (float, y, y)
+                                   (float, z, z) (float, intensity, intensity)
+                                   (float, roll, roll) (float, pitch, pitch) (float, yaw, yaw)
+                                   (double, time, time))
+
+typedef PointXYZIRPYT  PointTypePose;
+
+// When reading a point cloud, the frame is stored in the transformin array, and this function converts it into a readable bitmap in GTSAM.
+gtsam::Pose3 trans2gtsamPose(float transformIn[]){
+    return gtsam::Pose3(gtsam::Rot3::RzRyRx(transformIn[0], transformIn[1], transformIn[2]),
+                            gtsam::Point3(transformIn[3],transformIn[4],transformIn[5]));
+}
+
+//Converts a PointTypePose type point cloud to a 3D matrix, which is used to calculate angles when keyframes are selected.
+Eigen::Affine3f pclPointToAffine3f(PointTypePose thisPoint)
+{ 
+    return pcl::getTransformation(thisPoint.x, thisPoint.y, thisPoint.z, thisPoint.roll, thisPoint.pitch, thisPoint.yaw);
+}
+
+// Convert PointTypePose type point cloud to a bit pose that can be read in GTSAM.
+gtsam::Pose3 pclPointTogtsamPose3(PointTypePose thisPoint)
+{
+    return gtsam::Pose3(gtsam::Rot3::RzRyRx(double(thisPoint.roll), double(thisPoint.pitch), double(thisPoint.yaw)),
+                                gtsam::Point3(double(thisPoint.x),    double(thisPoint.y),     double(thisPoint.z)));
+}
+
+//Judging keyframes when the straight line distance is greater than 1 metre
+float surroundingkeyframeAddingDistThreshold = 1.0;
+// Judgement of keyframes when the angle is greater than 0.2 degrees
+float surroundingkeyframeAddingAngleThreshold = 0.2; 
+// Array of temporary point clouds
+float transformTobeMapped[6];
+//Searching for loopbacks requires greater than 10 seconds on both frames
+float historyKeyframeSearchRadius = 10.0;
+//Record radar time stamps
+ros::Time timeLaserInfoStamp;
+//convert timestamp to second
+double timeLaserInfoCur;
+bool aLoopIsClosed = false;
+// Construct an empty non-linear factor graph
+NonlinearFactorGraph gtSAMgraph;  
+// Declare the initial values of the factor graph and the optimisation results
+Values initialEstimate; 
+
+/*
+The factor graph only models the history of SLAM poses and the relationship between inputs and observations; 
+how to solve this factor graph, i.e., how to set the variables in such a way that the whole graph best meets all the constraints (with minimum error) 
+requires the use of an optimizer. In addition to the most common Gaussian-Newton and Levenberg-Marquardt optimizers for solving nonlinear problems, 
+GTSAM implements two incremental optimizers, iSAM,iSAM2*/
+ISAM2 *isam; 
```

Comment:
It's a bit hard to read the code as there are many globals. Hence, it's hard to see in a function whether a variable is local or global. This is is also an issue in some of the otehr examples, TBH. I typically deal with this in one of two ways:
- append g to all globals, e.g., gISAM
- have an Algorithm class, and make all of these instance variables: isam_.

Reply:
Here I used your first suggestion, since this is just a small test code, I won't wrap a separate class for them to make it easier to read. For a larger project I would have used the second, more robust approach.